diff --git i/google/appengine/tools/devappserver2/wsgi_server.py w/google/appengine/tools/devappserver2/wsgi_server.py
index c1729fc..b34ead7 100644
--- i/google/appengine/tools/devappserver2/wsgi_server.py
+++ w/google/appengine/tools/devappserver2/wsgi_server.py
@@ -197,7 +197,7 @@ class _SingleAddressWsgiServer(wsgiserver.CherryPyWSGIServer):
     host, port = self.bind_addr
     try:
       info = socket.getaddrinfo(host, port, socket.AF_UNSPEC,
-                                socket.SOCK_STREAM, 0, socket.AI_PASSIVE)
+                                socket.SOCK_STREAM, 1, socket.AI_PASSIVE)
     except socket.gaierror:
       if ':' in host:
         info = [(socket.AF_INET6, socket.SOCK_STREAM, 0, '',
