diff --git i/google/appengine/tools/devappserver2/watcher_common.py w/google/appengine/tools/devappserver2/watcher_common.py
index 26c2f52..074d551 100644
--- i/google/appengine/tools/devappserver2/watcher_common.py
+++ w/google/appengine/tools/devappserver2/watcher_common.py
@@ -18,7 +18,8 @@


 # Directories that we should not watch at all.
-_IGNORED_DIRS = ('.git', '.hg', '.svn')
+# 'genfiles' is a Khan Academy-specific build artifact directory.
+_IGNORED_DIRS = ('.git', '.hg', '.svn', 'genfiles')
 # File suffixes that should be ignored.
 _IGNORED_FILE_SUFFIXES = (
     # Python temporaries
